t/Mailbox/query/basic.t

```diff
@@ -0,0 +1,485 @@
+  my ($self) = @_;
+
+  plan skip_all => "Cyrus requires at least one mailbox"
+    if $self->server->isa('JMAP::TestSuite::ServerAdapter::Cyrus');
+
+  my $tester = $self->tester;
+
+  subtest "No arguments" => sub {
+    my $res = $tester->request({
+      using => [ "ietf:jmapmail" ],
+      methodCalls => [[
+        "Mailbox/query" => {},
+      ]],
+    });
+    ok($res->is_success, "Mailbox/query")
+      or diag explain $res->http_response->as_string;
+
+    jcmp_deeply(
+      $res->single_sentence("Mailbox/query")->arguments,
+      superhashof({
+        accountId => jstr($self->context->accountId),
+        state     => jstr(),
+        position  => jnum(0),
+        total     => jnum(0),
+        ids       => [],
+        canCalculateChanges => jbool(),
+      }),
+      "No mailboxes looks good",
+    ) or diag explain $res->as_stripped_triples;
+  };
+};
+
+# XXX - Test for basic response
+
+pristine_test "Mailbox/query filtering with filterConditions" => sub {
+  my ($self) = @_;
+
+  my $tester = $self->tester;
+
+  my $mailbox1 = $self->context->create_mailbox;
+
+  my $mailbox2 = $self->context->create_mailbox({
+    parentId => $mailbox1->id,
+  });
+
+  subtest "parentId" => sub {
+
+    subtest "does not have a parentId" => sub {
+      my $res = $tester->request({
+        using => [ "ietf:jmapmail" ],
+        methodCalls => [[
+          "Mailbox/query" => {
+            filter => {
+              parentId => undef,
+            },
+          },
+        ]],
+      });
+      ok($res->is_success, "Mailbox/query")
+        or diag explain $res->http_response->as_string;
+
+      my $ids = $res->single_sentence("Mailbox/query")->arguments->{ids};
+
+      jcmp_deeply($ids, supersetof($mailbox1->id), 'Got top-level mailbox');
+      jcmp_deeply($ids, noneof($mailbox2->id), 'Did not get sub-mailbox');
+    };
+
+    subtest "has a parentId" => sub {
+      my $res = $tester->request({
+        using => [ "ietf:jmapmail" ],
+        methodCalls => [[
+          "Mailbox/query" => {
+            filter => {
+              parentId => $mailbox1->id,
+            },
+          },
+        ]],
+      });
+      ok($res->is_success, "Mailbox/query")
+        or diag explain $res->http_response->as_string;
+
+      my $ids = $res->single_sentence("Mailbox/query")->arguments->{ids};
+
+      jcmp_deeply($ids, [ $mailbox2->id ], 'Got our one child mailbox');
+    };
+  };
+
+  subtest "hasRole" => sub {
+    # Find some mailboxes with roles
+    my $res = $tester->request({
+      using => [ "ietf:jmapmail" ],
+      methodCalls => [[
+        "Mailbox/get" => {},
+      ]],
+    });
+
+    my @with_roles = map {;
+      $_->{id}
+    } grep {;
+      $_->{role}
+    } @{ $res->single_sentence("Mailbox/get")->arguments->{list} };
+
+    plan skip_all => "No mailboxes with roles found, can't continue"
+      unless @with_roles;
+
+    subtest "false" => sub {
+      my $res = $tester->request({
```

Comment:
this pattern is just _everywhere_

Maybe we want a per-routine default-using, and then also a `$tester->reqsingle` that takes a single sentence to request, rather than the whole wrapper plus arrayref of triples.

Reply:
Sure